torchmultimodal/modules/encoders/text_encoder.py

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Callable, Optional
+
+import torch
+from torch import nn, Tensor
+from torchmultimodal.modules.layers.text_embedding import BERTTextEmbeddings
+from torchmultimodal.modules.layers.transformer import (
+    TransformerEncoder,
+    TransformerOutput,
+)
+from torchmultimodal.utils.attention import get_extended_attention_mask
+
+
+class TextEncoder(nn.Module):
+    """
+    General text transformer encoder with embeddings, similar to BERT / RoBERTa.
+    Can be constructed with any user-provided embeddings and encoder.
+
+    Based on https://github.com/huggingface/transformers/blob/main/src/transformers/models/bert/modeling_bert.py#L870
+
+    Attributes:
+        embeddings (nn.Module): Module that projects text token ids into embeddings.
+            ``forward()`` should follow interface:
+                input_ids: Optional[Tensor],
+                token_type_ids: Optional[Tensor],
+                position_ids: Optional[Tensor],
+                inputs_embeds: Optional[Tensor],
+        encoder (nn.Module): Module for transformer encoder. ``forward()`` should follow interface:
+            Inputs:
+                hidden_states: Tensor,
+                attention_mask: Optional[Tensor], shape [batch_size, num_heads, query_seq_length, key_seq_length]
+            Returns:
+                ``TransformerOutput``
```

Comment:
So I understand the motivation here, but I do see one potential drawback of this design down the road. We are gaining modularity by composing the text encoder out of embeddings and transformer encoder modules. At the same time, we are giving a pretty prescriptive API that users must follow to instantiate their own `TextEncoder` (specifically that the transformer encoder should return `TransformerOutput`). Right now we have a reasonable set of fields in that class, but if users want to do anything non-standard we will start getting requests to add a bunch of fields to TransformerOutput, which in the limit can look like [this](https://github.com/huggingface/transformers/blob/5a9957358cebd616e58b2d1ab3b887c2f2793b45/src/transformers/modeling_outputs.py#L24) (but in a single class). Btw I do not have a better suggestion at the moment, I am just here to nitpick 😃

Reply:
Yeah I think this is just a question of whether we want to use NamedTuples for outputs or keep it open and flexible. Having to update the fields (which can be done by the user in their PR itself) imo is much easier to deal with than having to refactor all callsites of TransformerEncoder when the output changes. I am not sure if there's any other alternative solution.